### DISCUSSION TITLE
Is IronScheme right for me?

### DISCUSSION THREAD

**Discussion by godotdot** [leppie/IronScheme#129] [Discussions] [answered]
Hello @leppie ,

This language looks very interesting but I not sure if it fits my needs or not; I'm hoping you might answer this overly-large list of questions I've prepared.

1 - Performance.
I came across these [benchmarks ](https://ecraven.github.io/r7rs-benchmarks/) and am wondering if the results seem accurate to you?
If not, can you explain why not?

Of course, benchmarks aren't always indicative of real-world performance, so the more important question might be "how does IronScheme's performance compare to C# or other .Net languages in real programs"?

Regarding my use case, do you think IronScheme performs well enough for game logic (not the actual engine)?

2 - Environments/Sandboxing
It appears that IronScheme has environments; I am under the impression that this allows restricting which functions can be called and which variables can be accessed.

I am interested in allowing users to create mods, but I'm not interested in allowing them to execute malicious code e.g. reading or writing arbitrary files.

Do you know if I am right in thinking IronScheme's environments are useful in preventing that sort of thing?

3 - Runtime modification of code
I am interested in being able to do stuff such as run my program, modify a file containing IronScheme code, and reevaluate it without having to restart the program.
I saw there were some nice examples of this [here](https://github.com/IronScheme/IronScheme/discussions/96).

It seems like quite a lot can be done, but are there any limitations I should be aware of?

4 - Debugging
I saw it mentioned that evaluated IronScheme code can be debugged:

>     enabling debug code generation (procedure)
> 
> Also R6RS's module system takes care of all (direct or indirect) library dependencies.
> 
> PS: If you enable debug code generation you can literally debug scheme scripts right in VS (it is not great, but not bad either).

Is there a procedure that can be called to enable this, or am I misunderstanding what was said?  Do you know if this would work in VSCode?

I think that about wraps up my questions; my apologies for asking so many!
Thank you for taking the time to look at all this.

**Comment by leppie** [comment, marked as answer]
Hi

1. IronScheme is aimed towards .NET compat and not really speed. If you want speed, look at Chez. 
2. It might be possible, but that is something you would have to do yourself. 
3. This is possible with the interaction environment, but not really modifiable, you would have to 'rebind' all the code to have the modifiable effect. EG you cant just redefine a used proc and have the behavior change. 
4. Source-level kinda works, but the .NET debugger is not very friendly for expression level stepping. It favours line/statement level stepping.  I dont know if it works in VSCode, but works in VS. I dont really use it that often, but did when writing the VS extension ( https://github.com/IronScheme/IronScheme.VisualStudio2/blob/master/IronScheme.VisualStudio/visualstudio.sls ).

Not sure if that works for you or not. Maybe others can chime in.

**Reply by godotdot** [reply]
Thank you for the quick reply!
